docs: add new driver (#174)
jina-ai/docs: chapters/extend/driver.rst

Development Guide: Add new Drivers
====================================

.. meta::
   :description: Development Guide: Add new Drivers
   :keywords: Jina, driver

.. note:: This guide assumes you have a basic understanding of Jina, if you haven't, please check out `Jina 101 <https://101.jina.ai>`_ first.

.. contents:: Table of Contents
    :depth: 2

Motivation
^^^^^^^^^^^

> As you might already learned from `Jina 101 <https://101.jina.ai>`_,

Reviewer: ```suggestion
As you might have already learned from `Jina 101 <https://101.jina.ai>`_,
```

a :term:`Driver` interprets incoming messages into :term:`Document` and extracts required fields for an :term:`Executor`.

> Jina already created `72 drivers <https://docs.jina.ai/chapters/all_driver/>`_ drivers inside Core,

Reviewer: I would avoid adding the number, so that we do not need to update this number constantly, can we say `several`?

and these `Drivers` should already fulfill most of the scenarios.

> However, when the existing Executors does not fit your specific use case,

Reviewer: ```suggestion
However, when the existing solutions do not fit your specific use case,
```

Development Guide: Add new Drivers
====================================

.. meta::

>    :description: Development Guide: Add new Drivers

Reviewer 2: Developer

   :keywords: Jina, driver

> .. note:: This guide assumes you have a basic understanding of Jina, if you haven't, please check out `Jina 101 <https://101.jina.ai>`_ first.

Reviewer 2: I actually think we should replace this with 
`.. warning::
    This is an advanced developer guide. The existing drivers should cover most use cases. If you add custom drivers, the Jina community may not be able to provide you with development support. `

Reviewer 2: I'm open to suggestions about the wording, but I want to make it clear that is not common. 

@bwanglzu @JoanFM what do you think?

Author: I like it @FionnD

.. note:: Before reading this guide, you might want to read `Development Guide: Add new Executors <../executor.rst>`_ first.

.. contents:: Table of Contents
    :depth: 2

Motivation
^^^^^^^^^^^

As you might have already learned from `Jina 101 <https://101.jina.ai>`_,
a :term:`Driver` interprets incoming messages into :term:`Document` and extracts required fields for an :term:`Executor`.

> Jina already created `several Drivers <https://docs.jina.ai/chapters/all_driver/>`_ Drivers inside Core,

Reviewer 2: this comes out as driver Drivers inside Core?

Author: I'll rephrase to Jina Core :)

Author: @FionnD I don't understand your comment, can you elaborate a bit on this?

Reviewer 2: See my slack message

and these `Drivers` should already fulfill most of the scenarios.
However, when the existing solutions do not fit your specific use case,

> you might be curious on how to **extend** Jina.

Reviewer 2: remove **on**

For instance, if you have created a customised `Executor`,

> and figure out the existent `Drivers` can not fit for your customised `Executor`,

Reviewer 2: Please change all to customized. We use the American spelling.

you might want to create a customised `Driver`.

Overview
^^^^^^^^^

> To make an extension of a Jina `Driver`, please follow the steps listed below:

Reviewer 2: There are two steps to adding an extension of a Jina `Driver`:

1. Decide which `Driver` class to inherit from.
2. Overwrite the **Core** method of the `Driver`.

Implementation
^^^^^^^^^^^^^^^

.. list-table:: Built-in Driver to Inherit
   :widths: 25 25 50 25
   :header-rows: 1

   * - Name
     - Corresponded Executor
     - Description
     - Core method
   * - `BaseEncodeDriver`
     - `BaseEncoder`
     - Driver bind with :meth:`encode` in the Executor.
     - `_apply_all`
   * - `BaseIndexDriver`
     - `BaseIndexer`
     - Driver bind with :meth:`add` in the Executor.
     - `_apply_all`
   * - `BaseSearchDriver`
     - `BaseIndexer`
     - Driver bind with :meth:`query` in the Executor.
     - `_apply_all`
   * - `BasePredictDriver`
     - `BaseClassifier`
     - Driver bind with :meth:`predict` in the Executor.
     - `_apply_all`
   * - `BaseLabelPredictDriver`
     - `BaseClassifier`
     - Driver bind with :meth:`predict` for label prediction.
     - `prediction2label`
   * - `BaseEvaluateDriver`
     - `BaseEvaluator`
     - Driver bind with :meth:`evaluate` in the Executor.
     - `extract`

> This is to say, if you want to create a customized `Executor` and it's associated `Driver`,

Reviewer 2: "This is to say". I would remove this, it's not useful.

> This is to say, if you want to create a customized `Executor` and it's associated `Driver`,

Reviewer 2: I would put this paragraph before the table, so that a user understands what they are reading

Author: okay!

> Follow the table above to decide which `BaseDriver` class to inherit from.

Reviewer 2: Lowercase F

Author: good catch!

After creating your customised `Driver` class, you need to implement your own Core method based on your specific need.

Customize Driver in Action: `MultimodalDriver`
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

We created `MultimodalDriver <https://github.com/jina-ai/jina/blob/master/jina/drivers/multimodal.py>`_ to better support Cross/Multi-modal search.
Assume a `Document` has 2 child `Documents` with different modalities, e.g. text and image.
The objective of `MultimodalDriver` is to extract embeddings from both modalities.
Our expected input and output can be represented as:

.. highlight:: shell
.. code-block:: shell

    Input:
    document:
            |- child document: {modality: mode1}
            |
            |- child document: {modality: mode2}
    Output:
    document: (embedding: multimodal encoding)
            |- child document: {modality: mode1}
            |
            |- child document: {modality: mode2}

> In the code snippt below, you should be able to see the logic of how we implemented the Driver.

Reviewer 2: snippet

Since class `MultiModalDriver` is based on `BaseEncodeDriver`, the :meth:`exec_fn` will be bind to the :meth:`encode` method,
to encode data from different modalities into embeddings.
Last but not least, we assign the `embeddings` property of each Document as the encoded vector representation.

.. highlight:: python
.. code-block:: python

    class MultiModalDriver(FlatRecursiveMixin, BaseEncodeDriver):
    """Extract multimodal embeddings from different modalities."""

        @property
        def positional_modality(self) -> List[str]:
            return self._exec.positional_modality

        def _get_executor_input_arguments(
            self, content_by_modality: Dict[str, 'np.ndarray']
        ) -> List['np.ndarray']:
            """From a dictionary ``content_by_modality`` it returns the arguments in the proper order so that they can be
            passed to the executor.
             :param content_by_modality: a dictionary of `Document content` by modality name
             :return: list of input arguments as np arrays
            """
            return [content_by_modality[modality] for modality in self.positional_modality]

>         def _apply_all(self, docs: 'DocumentSet', *args, **kwargs) -> None:

Reviewer: I think we do not need to duplicate exactly the code, we should just add small amount of code that can explain the concept

Author: @JoanFM I simplified the code
